core/store/src/trie/mem/flexible_data/value.rs

```diff
@@ -0,0 +1,102 @@
+use near_primitives::hash::CryptoHash;
+use near_primitives::state::{FlatStateValue, ValueRef};
+
+use super::FlexibleDataHeader;
+
+/// Flexibly-sized data header for a trie value, representing either an inline
+/// value, or a reference to a value stored in the State column.
+///
+/// The flexible part of the data is either the inlined value as a byte array,
+/// or a CryptoHash representing the reference hash.
+#[repr(C, packed(1))]
+#[derive(Clone, Copy)]
+pub struct EncodedValueHeader {
+    // The high bit is 1 if the value is inlined, 0 if it is a reference.
+    // The lower bits are the length of the value.
+    length_and_inlined: u32,
+}
+
+impl EncodedValueHeader {
+    const INLINED_MASK: u32 = 0x80000000;
+
+    fn decode(&self) -> (u32, bool) {
+        (
+            self.length_and_inlined & !Self::INLINED_MASK,
+            self.length_and_inlined & Self::INLINED_MASK != 0,
+        )
+    }
+}
+
+impl FlexibleDataHeader for EncodedValueHeader {
+    type InputData = FlatStateValue;
+    type View<'a> = ValueView<'a>;
+
+    fn from_input(value: &FlatStateValue) -> Self {
+        match value {
+            FlatStateValue::Ref(value_ref) => {
+                EncodedValueHeader { length_and_inlined: value_ref.length }
+            }
+            FlatStateValue::Inlined(v) => {
+                assert!(v.len() as u32 & Self::INLINED_MASK == 0);
```

Comment:
Same, `assert!(v.len() < INLINED_MASK as usize)`, to avoid issues with `v.len() > u32::MAX`, however unlikely that is.